Add wait-change API endpoint and client function

benhoyt opened the pull request:
This adds an endpoint to the Pebble API to wait for a change to be ready, with an optional timeout. It's intended for use with the upcoming "exec" feature ([spec](https://docs.google.com/document/d/1npbNQMokDUoGX1UT61CvnMq6lIJ6MnU6CwExkymHEcA/edit#) and [WIP PR](https://github.com/canonical/pebble/pull/61)). The LXD code has a similar feature, except changes are called "operations" in LXD, and there's a "wait operation" API endpoint.

This is particularly important for "pebble exec", so we can wait for the command to be finished without polling every short interval, and we can return a response immediately when the command finishes executing.

In future we could also use this new endpoint in the `pebble start` and `pebble stop` commands (and Python Operator Framework equivalents) to avoid polling there too.

hpidcock commented on internal/daemon/api_changes.go:
```diff
@@ -165,18 +165,52 @@ func v1GetChanges(c *Command, r *http.Request, _ *userState) Response {
+		}
+	} else {
+		// No timeout, wait indefinitely for change to be ready.
+		<-change.Ready()
```
select with wait on r.Context().Done()

benhoyt replied:
Thanks - fixed!

hpidcock commented on internal/daemon/api_changes.go:
```diff
@@ -165,18 +165,52 @@ func v1GetChanges(c *Command, r *http.Request, _ *userState) Response {
+		select {
+		case <-change.Ready():
+			timer.Stop() // change ready, release timer resources
+		case <-timer.C:
```
add select clause for select on r.Context().Done()

benhoyt replied:
Thanks - fixed!

hpidcock reviewed (approved):
Looks good, just two selects that need fixing.

benhoyt commented on client/changes.go:
```diff
@@ -157,3 +157,31 @@ func (client *Client) Changes(opts *ChangesOptions) ([]*Change, error) {
+		query.Set("timeout", opts.Timeout.String())
+	}
+
+	_, err := client.doSync("GET", "/v1/changes/"+id+"/wait", query, nil, nil, &chgd)
```
Good question. I had observed it to be a very long timeout / no timeout in my manual testing, but just confirmed that now in the code -- we always create the `http.Client` with no `Timeout` specified, so no timeout applies here. Arguably that's bad practice, but we can follow up in a subsequent PR if we want to change that.

benhoyt commented on client/changes.go:
```diff
@@ -157,3 +157,31 @@ func (client *Client) Changes(opts *ChangesOptions) ([]*Change, error) {
+	// If non-zero, wait at most this long before returning the current change
+	// data. The timeout elapsing is not considered an error, so if a timeout
+	// is specified, the caller should check Change.Ready to determine whether
+	// the change is actually finished.
```
Yeah, Harry and I went back and forth on that and ended up that it shouldn't be an error, primarily because when Pebble returns an error it can't also return other data (the Change in this case), and I thought it might be useful to have the current state of the change at the time of the timeout.

However, on reflection it makes for a less error-prone API if it's just an error, and in the success case the change is always ready. If the caller really wants the current state of a change in the case of a timeout/error, they can always make an additional call to `client.Change()` to get the current status. However, it's unlikely you'll want to.

In any case, I've used HTTP status code `504 Gateway Timeout` for this (along with a reasonable error message). It's not quite a perfect semantic match, but it seems close enough.

benhoyt commented on client/changes.go:
```diff
@@ -157,3 +157,31 @@ func (client *Client) Changes(opts *ChangesOptions) ([]*Change, error) {
+}
+
+// WaitChange waits for a given change to be finished (whether or not there
+// was an error, which is indicated by Change.Err being set).
```
I've updated this documentation.

benhoyt commented:
@niemeyer I've addressed your concerns -- mainly making the timeout an error and tweaking the doc comments. I'm going to go ahead and merge; we can always tweak later if necessary.
